_posts/2016-09-16-why-linting-errors-should-cause-tests-to-fail.md

```diff
@@ -0,0 +1,138 @@
+[ESLint](http://eslint.org/). This means that code that does not conform to our
+linting standards cannot be merged into master. While my colleague and I were
+both in agreement that linting is valuable, it was this change that sparked
+debate.
+
+## The simplified argument
+
+My colleague's main point against including ESLint in Travis was that, in
+a case where the site is down and we need to ship a hotfix, it would be
+terrible for a misaligned curly brace to prevent merging the fix and thus
+prolong our down-time.
+
+I must admit that I agree with his premise. It is far more important to bring
+the site back up than for master to be free of linting errors. However,
+I disagree that removing linting from our test suite is the best way to address
+this concern.
+
+I would argue that in the case of a site-down hotfix, you should have
+a mechanism for bypassing tests all together. There are many types of  of ways
+in which tests test failures which should not be a blocker to shipping
+a hotfix. For example:
+
+* Transient test failures cause by not properly mocking out the system clock
+* Test failures introduced by your package repository being down.
+
+I don't think anyone would suggest disabling tests all together just to avoid
+the possibility that one of these errors might block a hotfix.
+
+If you don't have a system that can work around those types of failures, you
+properly should. And if you _do_ have a system that can work around these
+failures, it would be appropriate to use it to work around a hypothetical
+misaligned curly brace.
+
+## Why I care so much
+
+By putting test failures on a pedestal above linting errors, I think we
+underestimate the value of linting, and the harm that linting errors can cause.
+
+Without enforcing your linting rules as part of your tests, you are guaranteed
+to end up with linting errors merged into your master branch. Let's explore
+three phases of a developer's workflow and see how they are negatively affected
+by this choice.
+
+### During code review
+
+If a reviewer can operate under the assumption that the code they are reviewing
+conforms to the project's linting rules, they don't need to manually lint the
+code they are reading.
+
+For example, if the reviewer sees that the author has removed a variable
+definition, they can trust that the author did not overlook a later usage of
+that variable. Without the linter's assurance, the reviewer would need to
+manually check the entire file for other usages of that variable.
+
+This applies to less overt errors as well. Did the author use the right number
+of spaces when they indented? If tests were configured to fail on linting
+errors, the reviewer wouldn't even need to ask herself that question, freeing
+her mental bandwidth to focus on other, more substantial, questions.
+
+### During deploy
+
+In addition to testing for issues of style, linters check for code validity. We
+should not deploy code that is invalid.
```

Comment:
I only mention this because most people view "linting" as separate from "invalid/incorrect/non-compiling" code, and will feel this doesn't make sense.